### ISSUE TITLE
FREEZE: Random bspwm freezes. bspc commands do not respond.

### ISSUE THREAD

**Issue by subnut** [baskerville/bspwm#1198]
This occurs TOTALLY randomly. 

NOT related to #1146, because `ps ax | grep bspwm` shows status `Ss`, not `T`. SIGCONT does not work.
NOT `sxhkd`. Keybinds work. Pressing keybinds that spawn `bspc` commands work. I can see the `bspc` processes piling up in the output of `ps ax | grep bspc`
NO external_rules configured by me. I use polybar though, and I don't know whether it does that. 
NOT same as the crash caused by `bspc subscribe --fifo`. REASON: when crashed like that, running `killall bspwm` only once is enough to restart bspwm. In my case it needs *twice*.

EDIT: I forgot to mention that when bspwm crashes, I can still interact with the active window with my keyboard. The mouse seems to vanish

**Comment by subnut** [comment]
@emanuele6 @baskerville Sorry for calling you out like that, but this is greatly affecting me. Mostly due to the cause that it occurs randomly, and I am unable to take any kind of precaution.

I'm sorry for calling out like this, but things become serious when you are in the middle of a meeting presenting something, and suddenly your window manager crashes. So, please help. I will be sharing any logs you want.

**Comment by baskerville** [comment]
You could:
- Use a debug binary (`make debug`).
- Enable core dumps.

**Comment by subnut** [comment]
> `make debug`
I am sorry, but I have never compiled any software before :sweat: 
As I am using a PKGBUILD, can you please show me the relevant parts to change in here (i.e. where exactly to put the `debug`) -
```
pkgver() {
  cd $pkgname
  git describe --long --tags | sed -r 's,^[^0-9]*,,;s,([^-]*-g),r\1,;s,[-_],.,g'
}

build() {
  cd $pkgname
  make PREFIX=/usr
}

package() {
  cd $pkgname
  make PREFIX=/usr DESTDIR="$pkgdir" install
  install -D -m644 LICENSE "${pkgdir}/usr/share/licenses/${pkgname}/LICENSE"
}
```

> Enable core dumps
Is this to be specifically enabled in bspwm? AFAIK, my coredumps are at `/var/lib/systemd/coredump`
But a quick `ls` didn't reveal any of bspwm
```
❯ ls /var/lib/systemd/coredump
core.my_zoom.1000.4072004182124398944245af06c79a26.1178774.1599493586000000.zst 
core.onedrive_tray.1000.4072004182124398944245af06c79a26.1351739.1599508764000000.zst 
core.onedrive_tray.1000.4072004182124398944245af06c79a26.1351749.1599508771000000.zst 
core.onedrive_tray.1000.4072004182124398944245af06c79a26.1351759.1599508779000000.zst 
core.onedrive_tray.1000.4072004182124398944245af06c79a26.1351769.1599508786000000.zst 
core.onedrive_tray.1000.4072004182124398944245af06c79a26.1351780.1599508794000000.zst 
core.onedrive_tray.1000.4072004182124398944245af06c79a26.1351825.1599508801000000.zst 
core.onedrive_tray.1000.4072004182124398944245af06c79a26.1351835.1599508809000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.1368.1599533696000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.274797.1599571235000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.274905.1599571242000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.274962.1599571250000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.275057.1599571257000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.275140.1599571265000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.276883.1599571323000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.277011.1599571331000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.361731.1599579849000000.zst 
core.picom.1000.4072004182124398944245af06c79a26.1307249.1599508756000000.zst 
core.picom.1000.4072004182124398944245af06c79a26.1351933.1599508847000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.227546.1599563485000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.234333.1599571227000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.275308.1599571298000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.277148.1599571475000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.278571.1599579842000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.361832.1599579858000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.5052.1599533756000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.661286.1599653380000000.zst 
core.polybar.1000.4072004182124398944245af06c79a26.1263587.1599508756000000.zst 
core.polybar.1000.4072004182124398944245af06c79a26.1263873.1599508756000000.zst 
core.polybar.1000.4072004182124398944245af06c79a26.1264317.1599508756000000.zst 
core.polybar.1000.4072004182124398944245af06c79a26.1285369.1599508756000000.zst 
core.polybar.1000.4072004182124398944245af06c79a26.1285780.1599508756000000.zst 
core.polybar.1000.4072004182124398944245af06c79a26.1286041.1599508756000000.zst 
core.polybar.1000.4072004182124398944245af06c79a26.1307113.1599508756000000.zst 
core.polybar.1000.4072004182124398944245af06c79a26.1307364.1599508756000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.12292.1599571227000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.227464.1599571227000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.227656.1599571227000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.234222.1599571227000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.234475.1599571227000000.zst 
```

I checked and there are files here older and newer than the most recent bspwm crash. Do I need to enable bspwm coredumps specifically?

**Comment by emanuele6** [comment]
> can you please show me the relevant parts to change in here (i.e. where exactly to put the debug)

```diff
 build() {
   cd $pkgname
-  make PREFIX=/usr debug
+  make PREFIX=/usr
 }
```

**Comment by subnut** [comment]
Okay ... I have put the `debug` where emanuele6 showed me to, and then installed bspwm. It is running fine right now.
But where should I look for logs, in case bspwm crashes again? Will it create a coredump automatically? Or do I need to run some command to do so?

----
Is there any way to restart bspwm when it crashes without losing all of my unsaved work?

---
Sorry if you were offended, Emanuele. In this scenario of online meetings, I had bspwm crash during important meetings _twice_. So, you see, the matter was becoming serious, and I needed immediately some way of creating logs for filing bug report here. As you are a regular contributor here, and Bastien was unavailable recently, so I thought it would be better to mention you too. Sorry again if you were annoyed.

---
Thank you both for your prompt responses

**Comment by subnut** [comment]
UPDATE: bspwm crashed again. I ran `killall bspwm && killall bspwm` (yes, twice) to stop it. Then logged in again from my Display Manager. 

I looked for coredumps related to bspwm, and there weren't any

See ...
```
❯ ls /var/lib/systemd/coredump
core.my_zoom.1000.4072004182124398944245af06c79a26.1178774.1599493586000000.zst 
core.onedrive_tray.1000.4072004182124398944245af06c79a26.1351739.1599508764000000.zst 
core.onedrive_tray.1000.4072004182124398944245af06c79a26.1351749.1599508771000000.zst 
core.onedrive_tray.1000.4072004182124398944245af06c79a26.1351759.1599508779000000.zst 
core.onedrive_tray.1000.4072004182124398944245af06c79a26.1351769.1599508786000000.zst 
core.onedrive_tray.1000.4072004182124398944245af06c79a26.1351780.1599508794000000.zst 
core.onedrive_tray.1000.4072004182124398944245af06c79a26.1351825.1599508801000000.zst 
core.onedrive_tray.1000.4072004182124398944245af06c79a26.1351835.1599508809000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.1368.1599533696000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.274797.1599571235000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.274905.1599571242000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.274962.1599571250000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.275057.1599571257000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.275140.1599571265000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.276883.1599571323000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.277011.1599571331000000.zst 
core.onedrive_tray.1000.d6f5a39061f740ecb81ab01f488269ce.361731.1599579849000000.zst 
core.picom.1000.4072004182124398944245af06c79a26.1307249.1599508756000000.zst 
core.picom.1000.4072004182124398944245af06c79a26.1351933.1599508847000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.227546.1599563485000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.234333.1599571227000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.275308.1599571298000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.277148.1599571475000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.278571.1599579842000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.361832.1599579858000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.5052.1599533756000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.661286.1599653380000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.857600.1599706446000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.959881.1599726799000000.zst 
core.picom.1000.d6f5a39061f740ecb81ab01f488269ce.989229.1599726824000000.zst 
core.polybar.1000.4072004182124398944245af06c79a26.1263587.1599508756000000.zst 
core.polybar.1000.4072004182124398944245af06c79a26.1263873.1599508756000000.zst 
core.polybar.1000.4072004182124398944245af06c79a26.1264317.1599508756000000.zst 
core.polybar.1000.4072004182124398944245af06c79a26.1285369.1599508756000000.zst 
core.polybar.1000.4072004182124398944245af06c79a26.1285780.1599508756000000.zst 
core.polybar.1000.4072004182124398944245af06c79a26.1286041.1599508756000000.zst 
core.polybar.1000.4072004182124398944245af06c79a26.1307113.1599508756000000.zst 
core.polybar.1000.4072004182124398944245af06c79a26.1307364.1599508756000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.12292.1599571227000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.227464.1599571227000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.227656.1599571227000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.234222.1599571227000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.234475.1599571227000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.814855.1599706446000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.815498.1599706446000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.815753.1599706446000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.815961.1599706446000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.816204.1599706446000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.816412.1599706446000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.816611.1599706446000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.972996.1599726427000000.zst 
core.polybar.1000.d6f5a39061f740ecb81ab01f488269ce.985779.1599726583000000.zst 
```

No entry related to bspwm!

**Comment by baskerville** [comment]
I think I might have been misled by the inappropriate use of the term *crash*.

Since *bspwm* is still running when the problem occurs, you could try to attach to the *bspwm* process using *GDB* and get a backtrace.

**Comment by subnut** [comment]
> you could try to attach to the bspwm process using GDB and get a backtrace

Sure! Could you please tell me the {syntax, commands to run}?

**Comment by SeerLite** [comment]
Hey @subnut! It's `gdb` not `gdm`.
Here are some resources on how to use it:
* [Attaching a process](https://stackoverflow.com/a/14371557/6791873)
* [Getting the backtrace](https://stackoverflow.com/a/4292614/6791873)

I'm linking these because I, too, didn't know how to use it and learned it from there. I know what I'm about to say might sound a bit rude, but you should start out by searching for stuff you don't understand when given these instructions by devs/maintainers, and only fall back to asking them when you can't find anything.
Most are already pretty busy maintaining the project or dealing with other issues, and it'd be really tiresome for them if they had to explain everything to each newbie that comes in with rare issues, you know?
 
> What do you suggest this should be called? Glitch?

I think "freezing" is fine ;)

Also important: What version of bspwm are you running? Maybe try installing the latest version from [Releases](https://github.com/baskerville/bspwm/releases)? Usually, you just download the source code, extract it, and run `make` followed by `make install` inside the folder.

**Comment by subnut** [comment]
@SeerLite Thank you very much! Actually, I had done a quick google search
about how to attach and get a backtrace using `gdb`, but could find no
related results. (Maybe I needed to search them separately.... :shrug:)

I shall definitely check out the links you mentioned.

> I think "freezing" is fine ;)

Agreed! :smile:

> What version of bspwm are you running?

I was running the latest release of bspwm at the time of filing this
issue. (I had installed it from the Arch repo).

But currently I am on ...
```
❯ bspwm -v
0.9.10-9-g853bb73
```
...compiled with the `debug` flag as suggested by baskerville in the comments above.

I had been using my laptop a little less recently, so I have not witnessed
any new Freezes. But who knows, when it might freeze again .....

When it does freeze, I shall definitely report back with the stack trace.
Thanks for helping!
